### PR TITLE
mds: make purge_queue delete objects asynchronously and keep accepting pushes

### PR DESCRIPTION
**Pull request by gmayyyha**
Signed-off-by: Yanhu Cao <gmayyyha@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by gmayyyha** [comment]
@batrick @ukernel Please help to review.

**Comment by batrick** [comment]
@kotreshhr please review this.

**Comment by kotreshhr** [comment]
> @kotreshhr please review this.

Sure @batrick I will take a look at it.

**Review by kotreshhr — COMMENTED** [review]
how is it determined that the objecter has slow requests ? The PR looks good to me otherwise.

**Comment by gmayyyha** [comment]
> how is it determined that the objecter has slow requests ? The PR looks good to me otherwise.

 Under high load, storing data to osd may take a long time.

File: src/mds/PurgeQueue.cc [hunk 1]
```diff
@@ -497,6 +497,99 @@ bool PurgeQueue::_consume()
   return could_consume;
 }
 
+class C_IO_PurgeItem_Commit : public Context {
+public:
+  C_IO_PurgeItem_Commit(PurgeQueue *pq, std::vector<PurgeItemCommitOp> &ops, uint64_t expire_to)
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:502]:
```suggestion
  C_IO_PurgeItem_Commit(PurgeQueue *pq, std::vector<PurgeItemCommitOp> ops, uint64_t expire_to)
```

File: src/mds/PurgeQueue.cc [hunk 2]
```diff
@@ -497,6 +497,99 @@ bool PurgeQueue::_consume()
...
+    : purge_queue(pq), expire_to(expire_to) {
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:503]:
```suggestion
    : purge_queue(pq), ops_vec(std::move(ops), expire_to(expire_to) {
```

File: src/mds/PurgeQueue.cc [hunk 3]
```diff
@@ -497,6 +497,99 @@ bool PurgeQueue::_consume()
...
+    ops_vec.swap(ops);
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:504]:
```suggestion
```

File: src/mds/PurgeQueue.cc [hunk 4]
```diff
@@ -514,38 +607,34 @@ void PurgeQueue::_execute_item(
   ops_high_water = std::max(ops_high_water, ops_in_flight);
   logger->set(l_pq_executing_ops_high_water, ops_high_water);
 
-  SnapContext nullsnapc;
+  std::vector<PurgeItemCommitOp> ops_vec;
+  auto submit_ops = [&]() {
+    finisher.queue(new C_IO_PurgeItem_Commit(this, ops_vec, expire_to));
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:612]:
```suggestion
    finisher.queue(new C_IO_PurgeItem_Commit(this, std::move(ops_vec), expire_to));
```

File: src/mds/PurgeQueue.h [hunk 5]
```diff
@@ -131,6 +153,8 @@ class PurgeQueue
   // to the queue (there is no callback for when it is executed)
   void push(const PurgeItem &pi, Context *completion);
 
+  void _commit_ops(int r, std::vector<PurgeItemCommitOp> &ops_vec, uint64_t expire_to);
```
Comment (batrick) [inline src/mds/PurgeQueue.h:156]:
```suggestion
  void _commit_ops(int r, const std::vector<PurgeItemCommitOp>& ops_vec, uint64_t expire_to);
```

File: src/mds/PurgeQueue.cc [hunk 6]
```diff
@@ -497,6 +497,99 @@ bool PurgeQueue::_consume()
...
+    purge_queue->_commit_ops(r, ops_vec, expire_to);
+  }
+
+private:
+  PurgeQueue *purge_queue;
+  std::vector<PurgeItemCommitOp> ops_vec;
+  uint64_t expire_to;
+};
+
+void PurgeQueue::_commit_ops(int r, std::vector<PurgeItemCommitOp> &ops_vec, uint64_t expire_to)
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:517]:
```suggestion
void PurgeQueue::_commit_ops(int r, const std::vector<PurgeItemCommitOp>& ops_vec, uint64_t expire_to)
```

File: src/mds/PurgeQueue.cc [hunk 7]
```diff
@@ -497,6 +497,99 @@ bool PurgeQueue::_consume()
...
+        objecter->remove(op.oid, op.oloc, op.item.snapc,
+                         ceph::real_clock::now(), op.flags,
+                         gather.new_sub());
+      }
+    } else if (op.type == PurgeItemCommitOp::PURGE_OP_ZERO) {
+      filer.zero(op.item.ino, &op.item.layout, op.item.snapc,
+                 0, op.item.layout.object_size, ceph::real_clock::now(), 0, true,
+                 gather.new_sub());
+    } else {
+      derr << "Invalid purge op: " << op.type << dendl;
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:560]:
```suggestion
      derr << "Invalid purge op: " << op.type << dendl;
      ceph_abort();
```

**Review by batrick — CHANGES_REQUESTED** [review]

**Comment by gmayyyha** [comment]
@batrick all done.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-01-12_22:03:59-fs-wip-pdonnell-testing-20210112.170550-distro-basic-smithi/

**Pull request merged**